### PR TITLE
add mkldnn pooling backward

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37218 BFloat16: enable prepacked weights's inference
* #37215 BFloat16: add explicit dtype support for to_mkldnn and to_dense
* #37199 add mkldnn relu backward and reshape backward
* #37147 add mkldnn batch_norm backward
* **#37146 add mkldnn pooling backward**
* #36122 add mkldnn linear backward
* #36121 enable mkldnn conv2d backward to support mkldnn tensor input

Differential Revision: [D22440968](https://our.internmc.facebook.com/intern/diff/D22440968)

### CODE REVIEW

File: aten/src/ATen/native/mkldnn/Pooling.cpp [hunk 1]
```diff
@@ -226,10 +346,100 @@ Tensor& mkldnn_adaptive_avg_pool2d_out(
...
+    bool ceil_mode,
+    bool count_include_pad,
+    c10::optional<int64_t> divisor_override) {
+  TORCH_CHECK(false, "mkldnn_avg_pool2d_backward_out: in-place mkldnn operations are not supported yet");
+}
+
+Tensor mkldnn_adaptive_avg_pool2d_backward(
+    const Tensor& grad_output,
+    const Tensor& input) {
+  TORCH_CHECK(input.dim() == 4, "mkldnn_adaptive_avg_pool2d_backward: Expect 2D input");
```
Comment (albanD) [inline aten/src/ATen/native/mkldnn/Pooling.cpp:412]: nit: maybe make the error message slightly more explicit that a 4D Tensor is expected.

File: aten/src/ATen/native/mkldnn/Pooling.cpp [hunk 2]
```diff
@@ -226,10 +346,100 @@ Tensor& mkldnn_adaptive_avg_pool2d_out(
...
+
+  auto output_size_vec = grad_output.sizes();
+  std::vector<int64_t> kernel_size(input.dim() - 2);
+  for (size_t i = 2; i < input.dim(); ++i) {
+    auto s1 = input.size(i);
+    auto s2 = output_size_vec[i];
+    TORCH_CHECK(s2 != 0, "output size can not be zero");
+    TORCH_CHECK(
+        s1 % s2 == 0,
+        "input size is not divisible by the output size is not supported yet");
```
Comment (albanD) [inline aten/src/ATen/native/mkldnn/Pooling.cpp:422]: nit: this sentence might be clearer: "The given input size is not divisible by the output size. This is not supported yet for mkldnn Tensors".

File: aten/src/ATen/native/mkldnn/Pooling.cpp [hunk 3]
```diff
@@ -144,6 +192,78 @@ static Tensor _mkldnn_pool2d(
...
+        stride_vec,
+        padding_vec_l,
+        padding_vec_r,
+        dilation_vec,
+        true /* ceil_mode */);
+
+    // adjust padding until output sizes agree
+    bool all_equal = false;
+    std::vector<int64_t> output_sizes;
+    while (!all_equal) {
```
Comment (albanD) [inline aten/src/ATen/native/mkldnn/Pooling.cpp:229]: Given that all ops are calling here. How expensive is this? Should there be a warning in the docs that the ceil_mode might be inefficient for mkldnn Tensors?

File: aten/src/ATen/native/mkldnn/Pooling.cpp [hunk 4]
```diff
@@ -17,8 +17,7 @@ Tensor mkldnn_max_pool2d(
     IntArrayRef padding,
     IntArrayRef dilation,
     bool ceil_mode) {
-  AT_ERROR(
-      "mkldnn_max_pool2d: ATen not compiled with MKLDNN support");
+  TORCH_CHECK(false, "mkldnn_max_pool2d: ATen not compiled with MKLDNN support");
```
Comment (albanD) [inline aten/src/ATen/native/mkldnn/Pooling.cpp:20]: Thanks for taking the time to fix these.

**Review by albanD — APPROVED** [review]
Looks good, small comments only.

File: aten/src/ATen/native/mkldnn/Pooling.cpp (hunk 1 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/Pooling.cpp:412]: @albanD, the file use many **AT_ASSERTM**, any difference with **TORCH_CHECK**?  **AT_ASSERTM** will be deprecated, may can replace them.

Comment (albanD) [reply aten/src/ATen/native/mkldnn/Pooling.cpp]: Yes you can replace all the AT_ASSERT with TORCH_CHECK

**Pull request closed without merging**